### ISSUE TITLE
Why is BC offline training so slow? (Imitation learning)

### ISSUE THREAD

**Issue by oseq** [Unity-Technologies/ml-agents#1561]
Hello i huge problem with offline BC training, that i already mentioned in one of my comments.
Traning with BC offline is very slow 😞
Reaching 1000 setps takes around 5 minutes, while for the same configuration in RL i reach 1000 steps in 40s.
While training editor seems very laggy like it has been played in 1 FPS (i have target frame rate set to 30). It doesnt seems like problem with my computer because looking at hardware monitor it does not have big load. Setting time scale above 1 only makes it worse. For time scale 30 it seems like 0.2 FPS.
I am using CPU version of Tensorflow.
My current configuration is:
>  trainer: offline_bc
>     max_steps: 2.0e5
>     batch_size: 128
>     summary_freq: 1000
>     max_steps: 5.0e4
>     batches_per_epoch: 20
>     use_recurrent: false
>     hidden_units: 256
>     learning_rate: 5.0e-4
>     num_layers: 4
>     sequence_length: 128
>     memory_size: 1024
>     demo_path: ./UnitySDK/Assets/Demonstrations/file.demo

thanks in advance for Your help

**Comment by Nevak** [comment]
Have you tried to increase the decision frequency of your agents? 
The higher the value, the less often the agent will have to calculate a new decision so you probably will see a performance improvement.

![image](https://user-images.githubusercontent.com/14808257/50699149-4e117480-1047-11e9-9a14-2ead6105b679.png)

**Comment by oseq** [comment]
@Nevak Thanks it helped a bit, i guess BC algorithm is just less effective than PPO ant thats because it is taking more time form one to another step.

**Comment by awjuliani** [comment]
HI @oseq 

Offline BC works fundamentally differently from PPO. While the simulation happens slower, training is not any slower. This is because the network is being updated using the offline demonstration data many times every step of the simulation.

**Comment by oseq** [comment]
@awjuliani Thanks for your resonse. So basicly it hard to directlty compare this two metods in terms of trening steps. For me RL yelds much better resaults than IL (despite i am working on car controll, which is a problem that is suitablefor IL). I have a problem: in BC i algorithm reach certain point which is not satisfying and than results drops, becameing even worse and not recovering from it. Do you maybe have some advice for me?

Here is graph. Red one is RL and pink one is IL. Both in same environment, both for learning time: 2 hours.
[graph](https://i.imgur.com/Qx4riWM.jpg)

**Comment by awjuliani** [comment]
Hi @oseq 

Bad performance is BC may be due to a demonstration which is too small in size. Can I ask how many examples/episodes you collected for the training data?

**Comment by oseq** [comment]
@awjuliani  Yeah sure, i have set of 14 runs with is 8500 expieriences. 
My brain parameters are:

> Vector observation size: 60 (using ray perception)
> Num Stacked Vector obs: 3
> Vector Action Size: 2 [continuous]
> No visual obs. 

 So that may be to small set. I will try with bigger one. Do you have any suggestions?

Thank you, for your help 👍

**Comment by awjuliani** [comment]
For BC 8500 may still be too few. We are working on a new Imitation Learning algorithm that will be able to learn from many fewer examples, but unfortunately that won't be released for another couple months.

**Comment by oseq** [comment]
@awjuliani ok so i will provide bigger set. Good to know that you working on better algorithm, sadly i have to do my comparative thesis now. I'm keeping my fingers crossed for you guys :)

**Comment by hyounesy** [comment]
At the risk of sounding totally ignorant, I am still not quite clear why we need a simulation running (i.e. Unity scene) during BC training if we already have recorded training data? In other words, isn't the recorded [observation, action] enough to train our model within python/TF?

**Comment by BartMartner** [comment]
Same, hyounesy.

**Comment by whiteGhost333** [comment]
Experiencing the same thing with :
# of experiences : 39828
episodes 10
mean reward 23.3

**Comment by mmattar** [comment]
@hyounesy - this is a good point. I'm re-tagging this issue as a feature request because it makes sense.

**Comment by awjuliani** [comment]
Hi all,

The simulation in Offline BC trainer exists to provide a test-time measure of performance. 

We also have a new imitation learning method, GAIL, which performs much better when there are only small amounts of demonstration data. I would recommend taking a look at that.

Closing this issue for now.
